README.rst
==============
Django Godfrey
==============

Provides a WSGIHandler mixin for better handling of uncaught exceptions
particularly when working on API endpoints that use JSON payloads.

Provides a set of error views that look at the Accepts HTTP header and
return appropriately formatted responses. This way someone's API-using client
doesn't get a boatload of HTML for an HTTP 404.

:Code:          https://github.com/willkg/django-godfrey
:License:       Mozilla Public License v2; see LICENSE
:Issues:        https://github.com/willkg/django-godfrey/issues
:Documentation: None, yet--this is pretty prototypey


.. Note::

   This is very pre-alpha! I'm playing with this code to figure out the
   best way to accomplish what I need to accomplish while doing the
   least amount of hacky stuff.


Handling uncaught exceptions
============================

When ``DEBUG=True`` and an uncaught exception happens, Django will return a
stunning amount of HTML-formatted information in the HTTP Response. This
sucks when you're debugging API issues since you have to wade through the
HTML to find the information you want.

Godfrey tweaks that so that if the ``Accepts`` HTTP header includes
``application/json``, it tweaks the ``HttpRequest`` to look as if it
were an AJAX request which causes the debug info to come out as
text. I find it a lot easier to get back unformatted text because the
interfaces I'm using to test API things (e.g. curl, command line
utilities, Firefox dev tools, etc.) display that pretty well.

When ``DEBUG=False`` and an uncaught exception happens and Django is
configured to send email, Django will send an error email to the ``ADMINS``
which includes a repr of the WSGIRequest. However, when dealing with
HTTP POST payloads that are in JSON format (and probably other
formats), then the repr of the WSGIRequest doesn't show the payload at
all. It shows this::

  POST:<QueryDict {}>

or this::

  POST:<Could not parse>

Neither is helpful.

Godfrey tweaks the WSGIRequest by putting the raw POST payload in the
``HTTP_X_POSTDATA`` META variable so that it shows up in the error logs.

To use the mixin, in your ``wsgi.py`` file, do something like this::

  import django
  from django.core.handlers.wsgi import WSGIHandler

  from godfrey.wsgi_utils import BetterDebugMixin


  class MyWSGIHandler(BetterDebugMixin, WSGIHandler):
      pass


  def get_wsgi_application()
      django.setup()
      return MyWSGIHandler()


  application = get_wsgi_application()


.. Note::

   This is really hacky, but I didn't see another way to do this without
   rewriting/overriding a lot of Django error handling.

   I'm happy to entertain better ideas.

   

Error views
===========

By default, the Django error views for 400, 403, 404 and 500 all render
HTML. This sucks for API-users.

You can change the template to something different, but what you really
want is to return the appropriate format according to the HTTP ``Accepts``
header.

Godfrey has an alternative set of error handlers that will look at the
``Accepts`` header and return appropriately formatted things.

To use the views, in your url conf file, do this::

  from godfrey.error_handlers import setup_handlers()
  setup_handlers()


.. Note::

   This could use a lot of work, but I think the gist of it is good.


CHANGELOG
==========
Change Log
==========

v0.1
====

* Initial release.
* Created ``godfrey.wsgi_utils.BetterDebugMixin``.
* Created ``godfrey.error_handlers``.

.. Note::

   This is very alpha-quality prototype stuff.


godfrey/wsgi_utils.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
import logging

from django.conf import settings
from django.views import debug
from django.utils import six


class BetterDebugMixin(object):
    """Provides better debugging data

    Developing API endpoints and tired of wading through HTML for HTTP
    500 errors?

    Working on POST API debugging and not seeing the POST data show up in
    the error logs/emails?

    Then this mixin is for you!

    It:

    * spits out text rather than html when DEBUG = True (OMG! THANK YOU!)
    * adds a "HTTP_X_POSTBODY" META variable so you can see the raw post data
      in error emails

    Usage:

    Create a WSGIHandler class. For example::

        import django
        from django.core.handlers.wsgi import WSGIHandler

        from godfrey.wsgi_utils import BetterDebugMixin


        class MyWSGIHandler(BetterDebugMixin, WSGIHandler):
            pass


        def get_wsgi_application()
            django.setup()
            return MyWSGIHandler()


        application = get_wsgi_application()

    """
    def handle_uncaught_exception(self, request, resolver, exc_info):
        if settings.DEBUG_PROPAGATE_EXCEPTIONS:
            raise

        logger = logging.getLogger('django.request')

        # This should be "bytes" here, though in Python 2, that's a
        # str type.
        #
        # FIXME: This is probably broken with Python 3.
        postbody = getattr(request, 'body', '')
        try:
            # For string-ish data, we truncate and encode in utf-8.
            postbody = postbody[:10000].encode('utf-8')
        except UnicodeEncodeError:
            # For binary, we say, 'BINARY CONTENT'
            postbody = 'BINARY CONTENT'

        # The logger.error generates a record which can get handled by the
        # AdminEmailHandler. Overriding all that machinery seems crazy, so
        # we're instead going to shove it in the META section.
        request.META['HTTP_X_POST_BODY'] = postbody

        logger.error('Internal Server Error: %s', request.path,
            exc_info=exc_info,
            extra={
                'status_code': 500,
                'request': request
            }
        )

        if settings.DEBUG:
            # request.is_ajax() == True will push this into doing text
            # instead of html which is waaaaaayyy more useful from an
            # API perspective. So if the Accept header is anything other
            # than html, we'll say it's an ajax request to return text.
            if 'html' not in request.META.get('HTTP_ACCEPT', 'text/html'):
                request.META['HTTP_X_REQUESTED_WITH'] = 'XMLHttpRequest'
            return debug.technical_500_response(request, *exc_info)

        # If Http500 handler is not installed, re-raise last exception
        if resolver.urlconf_module is None:
            six.reraise(*exc_info)

        # Return an HttpResponse that displays a friendly error message.
        callback, param_dict = resolver.resolve500()
        return callback(request, **param_dict)


godfrey/error_handlers.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
from django.http import JsonResponse
from django.views.defaults import (
    bad_request,
    page_not_found,
    permission_denied,
    server_error
)


def handler_http_400(request, template_name='400.html'):
    """HTTP 400 error handler that understands Accept header"""

    accepts = request.META.get('HTTP_ACCEPT', 'text/html')
    if 'application/json' in accepts:
        return JsonResponse(
            status=400,
            content_type='application/json',
            data={'error': 'bad request'}
        )

    if 'text/html' in accepts:
        return bad_request(request, template_name=template_name)


def handler_http_403(request, template_name='403.html'):
    """HTTP 403 error handler that understands Accept header"""

    accepts = request.META.get('HTTP_ACCEPT', 'text/html')
    if 'application/json' in accepts:
        return JsonResponse(
            status=403,
            content_type='application/json',
            data={'error': 'permission denied'}
        )

    if 'text/html' in accepts:
        return permission_denied(request, template_name=template_name)


def handler_http_404(request, template_name='404.html'):
    """HTTP 404 error handler that understands Accept header"""

    accepts = request.META.get('HTTP_ACCEPT', 'text/html')
    if 'application/json' in accepts:
        return JsonResponse(
            status=404,
            content_type='application/json',
            data={'error': 'page not found'}
        )

    if 'text/html' in accepts:
        return page_not_found(request, template_name=template_name)


def handler_http_500(request, template_name='500.html'):
    """HTTP 500 error handler that understands Aceept header"""

    # FIXME: Figure out how to do mimetype dispatching better
    accepts = request.META.get('HTTP_ACCEPT', 'text/html')
    if 'application/json' in accepts:
        return JsonResponse(
            status=500,
            content_type='application/json',
            data={'error': 'server error'}
        )

    if 'text/html' in accepts:
        return server_error(request, template_name=template_name)


def setup_handlers():
    import django.conf.urls

    django.conf.urls.handler400 = 'godfrey.error_handlers.handler_http_400'
    django.conf.urls.handler403 = 'godfrey.error_handlers.handler_http_403'
    django.conf.urls.handler404 = 'godfrey.error_handlers.handler_http_404'
    django.conf.urls.handler500 = 'godfrey.error_handlers.handler_http_500'
